File: NodeDemo/src/services/metricsService.ts
/**
 * Tracks performance metrics for async vs blocking endpoints
 */

interface RequestMetric {
    responseTime: number;
    timestamp: number;
}

interface EndpointMetrics {
    totalRequests: number;
    responseTimes: number[];
    avgResponseTime: number;
    minResponseTime: number;
    maxResponseTime: number;
    p50: number;
    p95: number;
    p99: number;
}

export class MetricsService {
    private metrics: Map<string, RequestMetric[]> = new Map();

    recordRequest(endpoint: string, responseTimeMs: number): void {
        if (!this.metrics.has(endpoint)) {
            this.metrics.set(endpoint, []);
        }

        this.metrics.get(endpoint)!.push({
            responseTime: responseTimeMs,
            timestamp: Date.now()
        });
    }

    getMetrics(): Record<string, any> {
        const result: Record<string, EndpointMetrics> = {};

        for (const [endpoint, metrics] of this.metrics.entries()) {
            const times = metrics.map(m => m.responseTime).sort((a, b) => a - b);

            if (times.length > 0) {
                result[endpoint] = {
                    totalRequests: times.length,
                    responseTimes: times,
                    avgResponseTime: this.average(times),
                    minResponseTime: Math.min(...times),
                    maxResponseTime: Math.max(...times),
                    p50: this.percentile(times, 50),
                    p95: this.percentile(times, 95),
                    p99: this.percentile(times, 99)
                };
            }
        }

        return {
            timestamp: new Date().toISOString(),
            eventLoopLag: this.getEventLoopLag(),
            memoryUsage: process.memoryUsage(),
            endpoints: result
        };
    }

    reset(): void {
        this.metrics.clear();
    }

    private average(arr: number[]): number {
        return Math.round(arr.reduce((a, b) => a + b, 0) / arr.length * 100) / 100;
    }

    private percentile(sortedArr: number[], p: number): number {
        const index = Math.ceil(sortedArr.length * p / 100) - 1;
        return Math.round(sortedArr[Math.max(0, index)] * 100) / 100;
    }

    private getEventLoopLag(): number {
        // Simple event loop lag measurement
        const start = Date.now();
        setImmediate(() => {
            const lag = Date.now() - start;
            return lag;
        });
        return 0; // Simplified for this demo
    }
}

File: CSharpDemo/AsyncAwaitWebApi/Services/MetricsService.cs
using System.Collections.Concurrent;
using System.Diagnostics;

namespace AsyncAwaitWebApi.Services;

/// <summary>
/// Tracks performance metrics for async vs sync endpoints
/// </summary>
public class MetricsService
{
    private readonly ConcurrentDictionary<string, List<long>> _responseTimes = new();
    private readonly ConcurrentDictionary<string, int> _requestCounts = new();
    private readonly object _lock = new object();

    public void RecordRequest(string endpoint, long responseTimeMs)
    {
        _responseTimes.AddOrUpdate(
            endpoint,
            _ => new List<long> { responseTimeMs },
            (_, list) =>
            {
                lock (_lock)
                {
                    list.Add(responseTimeMs);
                    return list;
                }
            });

        _requestCounts.AddOrUpdate(endpoint, 1, (_, count) => count + 1);
    }

    public object GetMetrics()
    {
        var metrics = new Dictionary<string, object>();

        foreach (var endpoint in _responseTimes.Keys)
        {
            List<long> times;
            lock (_lock)
            {
                times = _responseTimes[endpoint].ToList();
            }

            if (times.Any())
            {
                times.Sort();
                metrics[endpoint] = new
                {
                    TotalRequests = _requestCounts[endpoint],
                    AvgResponseTime = times.Average(),
                    MinResponseTime = times.Min(),
                    MaxResponseTime = times.Max(),
                    P50 = GetPercentile(times, 50),
                    P95 = GetPercentile(times, 95),
                    P99 = GetPercentile(times, 99)
                };
            }
        }

        return new
        {
            Timestamp = DateTime.UtcNow,
            ThreadPoolInfo = new
            {
                AvailableWorkerThreads = GetAvailableWorkerThreads(),
                AvailableCompletionPortThreads = GetAvailableCompletionPortThreads()
            },
            Endpoints = metrics
        };
    }

    public void Reset()
    {
        _responseTimes.Clear();
        _requestCounts.Clear();
    }

    private long GetPercentile(List<long> sortedList, int percentile)
    {
        if (sortedList.Count == 0) return 0;
        int index = (int)Math.Ceiling(sortedList.Count * percentile / 100.0) - 1;
        index = Math.Max(0, Math.Min(index, sortedList.Count - 1));
        return sortedList[index];
    }

    private int GetAvailableWorkerThreads()
    {
        ThreadPool.GetAvailableThreads(out int workerThreads, out _);
        return workerThreads;
    }

    private int GetAvailableCompletionPortThreads()
    {
        ThreadPool.GetAvailableThreads(out _, out int completionPortThreads);
        return completionPortThreads;
    }
}
